int32 num
float32 position_x
float32 position_y